rclcpp/include/rclcpp/experimental/intra_process_manager.hpp

```diff
@@ -390,28 +439,64 @@ class IntraProcessManager
       auto subscription_base = subscription_it->second.lock();
       if (subscription_base) {
         auto subscription = std::dynamic_pointer_cast<
-          rclcpp::experimental::SubscriptionIntraProcessBuffer<MessageT, Alloc, Deleter>
+          rclcpp::experimental::SubscriptionIntraProcessBuffer<PublishedType,
+          PublishedTypeAllocator, PublishedTypeDeleter, ROSMessageType>
           >(subscription_base);
         if (nullptr == subscription) {
-          throw std::runtime_error(
-                  "failed to dynamic cast SubscriptionIntraProcessBase to "
-                  "SubscriptionIntraProcessBuffer<MessageT, Alloc, Deleter>, which "
-                  "can happen when the publisher and subscription use different "
-                  "allocator types, which is not supported");
-        }
-
-        if (std::next(it) == subscription_ids.end()) {
-          // If this is the last subscription, give up ownership
-          subscription->provide_intra_process_message(std::move(message));
+          auto ros_message_subscription = std::dynamic_pointer_cast<
+            rclcpp::experimental::ROSMessageIntraProcessBuffer<ROSMessageType,
+            ROSMessageTypeAllocator, ROSMessageTypeDeleter>
+            >(subscription_base);
+
+          if (nullptr == ros_message_subscription) {
+            throw std::runtime_error(
+                    "--failed to dynamic cast SubscriptionIntraProcessBase to "
+                    "SubscriptionIntraProcessBuffer<MessageT, Alloc, Deleter>, which "
+                    "can happen when the publisher and subscription use different "
+                    "allocator types, which is not supported");
+          } else {
+            if constexpr (rclcpp::TypeAdapter<MessageT>::is_specialized::value) {
+              ROSMessageTypeAllocator ros_message_alloc(allocator);
+              auto ptr = ros_message_alloc.allocate(1);
+              ros_message_alloc.construct(ptr);
+              ROSMessageTypeDeleter deleter;
+              allocator::set_allocator_for_deleter(&deleter, &allocator);
+              rclcpp::TypeAdapter<MessageT>::convert_to_ros_message(*message, *ptr);
+              auto ros_msg = std::unique_ptr<ROSMessageType, ROSMessageTypeDeleter>(ptr, deleter);
+              ros_message_subscription->provide_intra_process_message(std::move(ros_msg));
+            } else {
+              if constexpr (std::is_same<MessageT, ROSMessageType>::value) {
+                if (std::next(it) == subscription_ids.end()) {
```

Comment:
this code block is almost identically copied 3 times within the function.
it may be better to make it a lambda or a standalone function

Reply:
We're now down to 2 copies, and they aren't exactly the same.  I'm going to mark this as resolved for now, but feel free to comment and unresolve if you disagree with the resolution.